src/scripts/bttvIntegration.js

```diff
@@ -0,0 +1,137 @@
+    const modifiedTextContent = words.map(word => word === keyword ? imageHtmlString : word).join(' ');
+    return modifiedTextContent;
+}
+
+
+/**
+ * @param {Array} arrayOfInputs values to be tested 
+ * @returns {Boolean} true if all inputs provided are both truthy and strings, false if not
+ */
+function areAllTruthyStrings(arrayOfInputs)  { 
+    let areTruthyStrings = true;
+    if (arrayOfInputs.length === 0){ areTruthyStrings = false }
+    for (input of arrayOfInputs){ 
+        if (!input || typeof input !== "string"){ areTruthyStrings = false }
+    }
+    return areTruthyStrings
+}
+
+
+/**
+ * @param {string} username twitch username, case insensitive
+ * @returns {string} twitch user id
+ */
+async function getTwitchUserId(username, fetchFunction){ 
+    if (!username || typeof username !== "string"){ 
+        return null;
+    }
+    fetchFunction = fetchFunction || defaultFetch
+    const apiBaseUrl = "https://streamraiders.tips/_functions/getTwitchProfileData/";
+    const urlToQuery = apiBaseUrl + username;
+    const twitchUserData = await fetchFunction(urlToQuery);
+    return new Promise( resolve => { 
+        resolve(twitchUserData.id)
+    })
+}
+
+
+/**
+ * @param {string} userId twitch's user ID 
+ * @param {Function} fetchFunction OPTIONAL - a specific fetch function that returns a promise
+ * that resolves to a json of the fetch data.
+ * @returns {Object} key is the friendly emote name, value is the BTTV emote id
+ */
+async function getBttvChannelEmoteDict(userId, fetchFunction){
+    if ( !userId || typeof userId !== "string") {return null} 
+    fetchFunction = fetchFunction || defaultFetch
+    const apiBaseUrl = "https://api.betterttv.net/3/cached/users/twitch/"
+    const urlToQuery = apiBaseUrl + userId;
+    const bttvChannelData = await fetchFunction(urlToQuery);
+    const emoteDict = convertBttvChannelDataToEmoteDict(bttvChannelData);
+    return new Promise( resolve=> { 
+        resolve(emoteDict);
+    })
+}
+
+
+/**
+ * @param {string} url url to retrieve json from
+ * @returns {Object} json data received from fetch
+ */
+async function defaultFetch(url){ 
+    const response = await fetch(url);
+    const json = await response.json();
+    return new Promise(resolve => { 
+        resolve(json);
+    })
+}
+
+
+/**
+ * @param {Object} channelData  data structure provided by BTTV's api for channel queries
+ * @returns {Object} key is the friendly emote name, value is the BTTV emote id
+ */
+function convertBttvChannelDataToEmoteDict(channelData){ 
+    if (!channelData || typeof channelData !== "object") { return null };
+    let emotesArray = []
+    if (channelData.channelEmotes) { emotesArray.push(...channelData.channelEmotes) }
+    if (channelData.sharedEmotes) { emotesArray.push(...channelData.sharedEmotes)}
+    let emotesDict = {};
+    for (emote of emotesArray){ 
+        const emoteName = emote.code;
+        emotesDict[emoteName] = emote.id;
+    }
+    return emotesDict
```

Comment:
It seems to me like we could cut out the middleman of the array and skip straight to assembling `emotesDict`.

```suggestion
    let emotesDict = {};
    if (channelData.channelEmotes) {
        channelData.channelEmotes.forEach(emote => emoteDict[emote.code] = emote.id;
    }
    if (channelData.sharedEmotes) {
        channelData.sharedEmotes.forEach(emote => emoteDict[emote.code] = emote.id;
    }
    return emotesDict;
```

Thoughts?

Reply:
Hey, I LOVE this!! Great thinking!